### ISSUE TITLE
Targeted suppression for MA0011 analyzer

### ISSUE THREAD

**Issue by ygeerts** [meziantou/Meziantou.Analyzer#890]
### Description

We are using custom types (for example: Date, Time, ...).
The issue we're having is that the MA0011 analyzer always reports a diagnostic error as it does not have knowledge of whether the format is invariant or not for such a custom type.

### Examples

`_ = new System.Text.StringBuilder(10).Append($"{Date.Today:o}-{Time.Now:o}");`

In this code snippet we only have custom types so our suppressor knows exactly whether it should suppress the diagnostic error or not, based on our own list of formats which are considered invariant.

`_ = new System.Text.StringBuilder(10).Append($"{Date.Today:o}-{DateTime.Now:o}-{Time.Now:o}");`

In this code snippet we have one interpolated string which contains a mix of custom types (Date & Time) and a standard type (DateTime).
In this scenario, an MA0011 diagnostic error will be reported (because it doesn't know whether the 'o' format is invariant or not in our custom types).
In our suppressor, we can't really know if we can suppress the diagnostic error or not, because we don't know if the 'o' format for DateTime is considered invariant.

The issue is that we can't know for certain which formats the analyzer considers as invariant, and we only want to suppress the diagnostic error when all components of the interpolated string are considered invariant.
 
### Our solution

To resolve this, we are in the process of writing a suppressor to suppress MA0011 errors when we're using these custom types.
We know which formats are considered invariant when using these custom types, but we also need to know which formats the MA0011 analyzer considers invariant for BCL types.
 
We see the following options to implement this:

1. Duplicate the knowledge that MA0011 has concerning the formats of BCL types.
2. Expose the knowledge of these formats in Meziantou.Analyzer
3. Extend the MA0011 analyzer to also support custom types.

We're happy to provide a pull request if options 2 or 3 are viable for you.

Thanks!

**Comment by meziantou** [comment]
Currently, the only way to "configure" the rule is to add the [`Meziantou.Analyzer.Annotations.CultureInsensitiveTypeAttribute` attribute](https://github.com/meziantou/Meziantou.Analyzer/blob/main/src/Meziantou.Analyzer.Annotations/CultureInsensitiveTypeAttribute.cs) on a type. You can define the attribute in your project or reference the [Meziantou.Analyzer.Annotations package](https://www.nuget.org/packages/Meziantou.Analyzer.Annotations).

Note that it doesn't support providing which format are invariant. Is it something that would be needed in your use-case?

**Comment by meziantou** [comment]
The latest version allows specifying a list of culture invariant formats for any types. So, you should not need a custom suppressor.

```c#
// At assembly level
[assembly: CultureInsensitiveType(typeof(Custom), "o")]

// Per type
[CultureInsensitiveType("o")]
class Sample
{
}
```

**Comment by drieseng** [comment]
@meziantou, thanks a lot for the effort!!!

Could it be that you haven't yet published a new version of **Meziantou.Analyzer.Annotations**?
No big deal for us though, since we'll be defining **Meziantou.Analyzer.Annotations.CultureInsensitiveTypeAttribute* locally to avoid adding **Meziantou.Analyzer.Annotations** as a dependency to our library.

I had a look at your implementation and noticed that it's not possible to indicate that a **null** format or the parameterless `ToString()` is invariant without considering the whole type (all formats) as being invariant.

Would it be possible to consider the *CultureInsensitiveTypeAttribute* without **format** argument as a means of saying all formats are invariant, and whenever the **format** argument is effectively specified only consider its value as invariant. A **null** value would mean that the **null** format is invariant, but how to indicate that the parameterless `ToString()` is invariant though.

By testing your changes, I also found that the parameterless `ToString()` is always considered invariant for custom types:

```c#
var t = new MyType(); // no use of the CultureInsensitiveType annotation
Console.WriteLine(t.ToString("o")); // MA0011
Console.WriteLine(t); // Nothing!
Console.WriteLine(t.ToString()); // Nothing!
Console.WriteLine(t.ToString(format: null)); // MA0011
```

For string interpolation, it's ok though:

```c#
var t = new MyType(); // no use of the CultureInsensitiveType annotation
Console.WriteLine($"{t:o}"); // MA0076
Console.WriteLine($"{t}"); // MA0076
Console.WriteLine($"{t:null}"); // MA0076
```

Let me know if you want a compilable repro for this issue.

**Comment by drieseng** [comment]
I also don't understand this [test](https://github.com/meziantou/Meziantou.Analyzer/blob/3d7df2e31637295e94ce664284e75a2378725e95/tests/Meziantou.Analyzer.Test/Rules/DoNotUseImplicitCultureSensitiveToStringAnalyzerTests.cs#L422-L438). Why you would consider a (custom) format for which there's no corresponding **CultureInsensitiveTypeAttribute** as invariant?

Perhaps (most likely) I'm misreading the test.

**Comment by drieseng** [comment]
One last remark:
Wouldn't it be better to cache the result of the invariant format check for the combination of type (symbol) and format.
It wouldn't surprise me if it would have a measurable cost in a large code base (like ours).
Of course, feel to ignore this remark unless I can show you the numbers :)

**Comment by meziantou** [comment]
> Could it be that you haven't yet published a new version of Meziantou.Analyzer.Annotations?

Good catch! It should be available in a few minutes

> No big deal for us though, since we'll be defining *Meziantou.Analyzer.Annotations.CultureInsensitiveTypeAttribute locally to avoid adding Meziantou.Analyzer.Annotations as a dependency to our library.

You can add a dependency on the lib. In the end, it should not be part of the dependencies of your NuGet packages as all the types are conditional (`[System.Diagnostics.Conditional("MEZIANTOU_ANALYZER_ANNOTATIONS")]`). So, the dependency should be elided.

> I also don't understand this [test](https://github.com/meziantou/Meziantou.Analyzer/blob/3d7df2e31637295e94ce664284e75a2378725e95/tests/Meziantou.Analyzer.Test/Rules/DoNotUseImplicitCultureSensitiveToStringAnalyzerTests.cs#L422-L438)

The format `abc` is culture-insensitive (attribute). However, `other` is not. So, it should report the diagnostic.

> Of course, feel to ignore this remark unless I can show you the numbers :)

You can get numbers using the binlog: https://www.meziantou.net/understanding-the-impact-of-roslyn-analyzers-on-the-build-time.htm

**Comment by drieseng** [comment]
>> I also don't understand this [test](https://github.com/meziantou/Meziantou.Analyzer/blob/3d7df2e31637295e94ce664284e75a2378725e95/tests/Meziantou.Analyzer.Test/Rules/DoNotUseImplicitCultureSensitiveToStringAnalyzerTests.cs#L422-L438)
> 
> The format `abc` is culture-insensitive (attribute). However, `other` is not. So, it should report the diagnostic.

The test doesn't make it clear to me that a diagnostic is expected.

[This](https://github.com/meziantou/Meziantou.Analyzer/blob/3d7df2e31637295e94ce664284e75a2378725e95/tests/Meziantou.Analyzer.Test/Rules/DoNotUseImplicitCultureSensitiveToStringAnalyzerTests.cs#L402-L419) test does **not** expect a diagnostic (since "abc" is defined as invariant) and I don't see how you express the fact that you don't expect a diagnostic here and do expect one in the other test.

Both do:

```c#
await CreateProjectBuilder()
              .WithSourceCode(sourceCode)
              .ValidateAsync();
```

Again, I must be missing something.

Note:
I also updated my earlier comment for the string interpolation example.
Also note that we've explicitly set **MA0011.exclude_tostring_methods** to **false**, but apparently did not set **MA0076.exclude_tostring_methods** (to **false**).  Is suppose **false** is the default for both though.

**Comment by meziantou** [comment]
> [This](https://github.com/meziantou/Meziantou.Analyzer/blob/3d7df2e31637295e94ce664284e75a2378725e95/tests/Meziantou.Analyzer.Test/Rules/DoNotUseImplicitCultureSensitiveToStringAnalyzerTests.cs#L402-L419) test does not expect a diagnostic (since "abc" is defined as invariant) and I don't see how you express the fact that you don't expect a diagnostic here and do expect one in the other test.

If you look at the source code, you can see that some part of the code is surrounded with a marker (`[|code|]`). This indicates where a diagnostic is expected. If there is no marker, then no diagnostic is expected.

Both tests have `
[assembly: CultureInsensitiveType(typeof(DateTime), "abc")]` meaning that the format `abc` is considered as invariant for DateTime type.

```
$"abc{new System.DateTime():abc}"; // should not report, as `abc` format is invariant
$"abc[|{new System.DateTime():other}|]"; // should report (see the markers) as `other` format is not invariant
```

Does it make sense?

**Comment by meziantou** [comment]
I've made some improvements:
- format now supports `null`
- `ToString` is more consistent as it reuses more logic with the other rules.

**Comment by drieseng** [comment]
> If you look at the source code, you can see that some part of the code is surrounded with a marker (`[|code|]`). This indicates where a diagnostic is expected. If there is no marker, then no diagnostic is expected.

I missed that subtility, sorry about that.

**Comment by drieseng** [comment]
> I've made some improvements:
> 
> * format now supports `null`
> * `ToString` is more consistent as it reuses more logic with the other rules.

You're making great process, thanks!!
I've submitted a few issues though. Makes it easier to track.

"You" still need to make the **format** argument nullable in **CultureInsensitiveTypeAttribute**. Let me know if you want me to submit a PR for this.

**Comment by drieseng** [comment]
I also wondered whether it would've been better not to report any "culture-related" diagnostic for custom types, except when **CultureInsensitiveTypeAttribute** annotations are applied for specific formats. All formats for which no **CultureInsensitiveTypeAttribute** is defined would then be considered culture-variant.

This would be considered a breaking change of course so you'd probably have to opt-in to this behaviour.

If (big if) you want to take this breaking change, then it's no longer useful to **CultureInsensitiveTypeAttribute** on a type without a format, and it would make more sense to rename it to **CultureInsensitiveFormatAttribute**.

**Example 1:**

```c#
using System;
using Meziantou.Analyzer.Annotations;

internal sealed class Program
{
    private static void Main(string[] args)
    {
        Foo foo = new Foo();

        // Do not report MA0011 or MA0076 for any of these since Meziantou has no idea which formats are culture variant and which not
        Console.WriteLine(foo.ToString());
        Console.WriteLine(foo.ToString(format: null));
        Console.WriteLine(foo.ToString(format: "D"));
        Console.WriteLine($"{foo:null}");
        Console.WriteLine($"{foo:D}");
        Console.WriteLine($"{foo}");
    }
}

public sealed class Foo : IFormattable
{
    public string ToString(string format)
    {
        return string.Empty;
    }

    public string ToString(IFormatProvider formatProvider)
    {
        return string.Empty;
    }

    public string ToString(string format, IFormatProvider formatProvider)
    {
        return string.Empty;
    }

    public override string ToString()
    {
        return string.Empty;
    }
}
```

**Example 2:**

```c#
using System;
using Meziantou.Analyzer.Annotations;

internal sealed class Program
{
    private static void Main(string[] args)
    {
        Foo foo = new Foo();

        Console.WriteLine(foo.ToString()); // no diagnostics as "null" is defined as invariant
        Console.WriteLine(foo.ToString(format: null)); // no diagnostics as "null" is defined as invariant
        Console.WriteLine(foo.ToString(format: "o")); // no diagnostics as "o" is defined as invariant
        Console.WriteLine(foo.ToString(format: "D")); // MA0011 since "D" is not defined as invariant
        Console.WriteLine($"{foo}"); // no diagnostics as "null" is defined as invariant
        Console.WriteLine($"{foo:null}"); // no diagnostics as "null" is defined as invariant
        Console.WriteLine($"{foo:o}"); // no diagnostics as "o" is defined as invariant
        Console.WriteLine($"{foo:D}"); // MA0076 since "D" is not defined as invariant
    }
}

[CultureInsensitiveType(format: null)]
[CultureInsensitiveType(format: "o")]
public sealed class Foo : IFormattable
{
    public string ToString(string format)
    {
        return string.Empty;
    }

    public string ToString(IFormatProvider formatProvider)
    {
        return string.Empty;
    }

    public string ToString(string format, IFormatProvider formatProvider)
    {
        return string.Empty;
    }

    public override string ToString()
    {
        return string.Empty;
    }
}
```

**Comment by meziantou** [comment]
> "You" still need to make the format argument nullable in CultureInsensitiveTypeAttribute. Let me know if you want me to submit a PR for this.

Fixed by b1826a70e0f448f3fb5b822a3c41aa7100238e94

**Comment by meziantou** [comment]
@drieseng Can we close this issue and continue the discussion in the new issues you opened?

**Comment by drieseng** [comment]
@meziantou Of course. Again a big thanks!
